modules/apps/portal-cache/portal-cache-impl/src/main/java/com/liferay/portal/cache/internal/dao/orm/CTAwarePortalCache.java

```diff
@@ -0,0 +1,207 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2024 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.portal.cache.internal.dao.orm;
+
+import com.liferay.portal.kernel.cache.MultiVMPool;
+import com.liferay.portal.kernel.cache.PortalCache;
+import com.liferay.portal.kernel.cache.PortalCacheListener;
+import com.liferay.portal.kernel.cache.PortalCacheListenerScope;
+import com.liferay.portal.kernel.cache.PortalCacheManager;
+import com.liferay.portal.kernel.change.tracking.CTCollectionThreadLocal;
+import com.liferay.portal.kernel.change.tracking.cache.CTCacheThreadLocal;
+import com.liferay.portal.kernel.model.change.tracking.CTModel;
+
+import java.io.Serializable;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * @author Tina Tian
+ */
+public class CTAwarePortalCache
+	implements PortalCache<Serializable, Serializable> {
+
+	public CTAwarePortalCache(
+		MultiVMPool multiVMPool, String portalCacheName, boolean mvcc,
+		boolean sharded) {
+
+		_multiVMPool = multiVMPool;
+		_portalCacheName = portalCacheName;
+		_mvcc = mvcc;
+		_sharded = sharded;
+
+		_productionPortalCache =
+			(PortalCache<Serializable, Serializable>)multiVMPool.getPortalCache(
+				portalCacheName, mvcc, sharded);
+	}
+
+	public void destroy() {
+		_multiVMPool.removePortalCache(
+			_productionPortalCache.getPortalCacheName());
+
+		for (PortalCache<Serializable, Serializable> ctPortalCache :
+				_ctPortalCaches.values()) {
+
+			_multiVMPool.removePortalCache(ctPortalCache.getPortalCacheName());
+		}
+
+		_ctPortalCaches.clear();
+	}
+
+	@Override
+	public Serializable get(Serializable key) {
+		if (CTCacheThreadLocal.isCTCacheEnabled()) {
+			PortalCache<Serializable, Serializable> ctPortalCache =
+				_getCTPortalCache(CTCollectionThreadLocal.getCTCollectionId());
+
+			return ctPortalCache.get(key);
+		}
+
+		return _productionPortalCache.get(key);
+	}
+
+	@Override
+	public List<Serializable> getKeys() {
+		throw new UnsupportedOperationException();
+	}
+
+	@Override
+	public PortalCacheManager<Serializable, Serializable>
+		getPortalCacheManager() {
+
+		return _productionPortalCache.getPortalCacheManager();
+	}
+
+	@Override
+	public String getPortalCacheName() {
+		return _portalCacheName;
+	}
+
+	@Override
+	public boolean isMVCC() {
+		return _mvcc;
+	}
+
+	@Override
+	public boolean isSharded() {
+		return _sharded;
+	}
+
+	@Override
+	public void put(Serializable key, Serializable value) {
+		put(key, value, PortalCache.DEFAULT_TIME_TO_LIVE);
+	}
+
+	@Override
+	public void put(Serializable key, Serializable value, int timeToLive) {
+		if (CTCacheThreadLocal.isCTCacheEnabled() &&
+			(value instanceof CTModel)) {
```

Comment:
this is always going to fail because value will be passing in the CachedModel which is not an instanceof CTModel